resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/kubernetes/KubernetesClusterSchedulerBackend.scala

```diff
@@ -185,23 +190,86 @@ private[spark] class KubernetesClusterSchedulerBackend(
 
   private val allocatorRunnable: Runnable = new Runnable {
 
+    private val MAX_EXECUTOR_LOST_REASON_CHECKS = 10
+    private val executorsToRecover = new mutable.HashSet[String]
+    // Maintains a map of executor id to count of checks performed to learn the loss reason
+    // for an executor.
+    private val executorReasonChecks = new mutable.HashMap[String, Int]
+
     override def run(): Unit = {
-      if (totalRegisteredExecutors.get() < runningExecutorPods.size) {
+      removeFailedExecutors()
+      if (totalRegisteredExecutors.get() < runningExecutorsToPods.size) {
```

Comment:
Should `runningExecutorsToPods.size` be guarded by `RUNNING_EXECUTOR_PODS_LOCK`? It probably should because all reads/writes to a shared variable should be guarded by the same lock if any read/write is guarded. And regardless of if it should or not, it's better to use a `val` to hold the value so we only call `size` once.